### PR TITLE
Bumped versions of k8s dependencies and controller-runtime and k8s dependencies

### PR DESCRIPTION
**Pull request by NikhilSharmaWe**
**What this PR does / why we need it**:

This PR bumps versions of k8s dependencies and controller-runtime and k8s dependencies to be compatible with https://github.com/kubernetes-sigs/kubebuilder.

### CODE REVIEW

File: go.mod
Comment (camilamacedo86) [reply go.mod:20]: You need to solve the breaking changes here.

Comment (NikhilSharmaWe) [reply go.mod:20]: @camilamacedo86 The error regarding the types used are showing in `k8s.io/kubectl` and `k8s.io/cli-runtime`.
So do we need to make changes there to fix the issue and then use the new tag/version of them instead?

Comment (camilamacedo86) [reply go.mod:20]: Yes. We need to fix the project to work with the latest versions. The error says what brokes:
```
want: github.com/google/gnostic/openapiv2
but
got: github.com/googleapis/gnostic/openapiv2
```
It means that where we use github.com/googleapis/gnostic/openapiv2 we need to change to use  github.com/google/gnostic/openapiv2 instead now.

File: go.mod
Comment (camilamacedo86) [inline go.mod:21]: We have a bug fix for c+r. we need to bump v0.12.1

File: go.mod
Comment (everettraven) [inline go.mod:19]: Looks like the bump in the kubectl package version has resulted in some test failures due to changes in the api.

Looks to be that the changes in the api affect these lines:
https://github.com/kubernetes-sigs/kubebuilder-declarative-pattern/blob/e5a105c6e17b2f267a5d79b130334029db898aad/pkg/patterns/declarative/pkg/applier/direct.go#L45

Has changed to the signature:
```go
 v, err := cmdutil.NewFactory(&genericclioptions.ConfigFlags{}).Validator(validationDirective string, verifier *resource.QueryParamVerifier)
```

It seems like the validationDirective options are: https://github.com/kubernetes/apimachinery/blob/97e5df2d0258ac077093867fabf508c1fc31f53b/pkg/apis/meta/v1/types.go#L515-L522

Here is more information on the `resources.QueryParamVerifier`: https://pkg.go.dev/k8s.io/cli-runtime/pkg/resource#QueryParamVerifier

It also seems like the `NewApplyOptions` function was removed, so this line will also need to be modified:
https://github.com/kubernetes-sigs/kubebuilder-declarative-pattern/blob/e5a105c6e17b2f267a5d79b130334029db898aad/pkg/patterns/declarative/pkg/applier/direct.go#L68

The godocs for the `apply` package that the `NewApplyOptions` function came from can be found here: https://pkg.go.dev/k8s.io/kubectl@v0.24.0/pkg/cmd/apply

**Comment by NikhilSharmaWe** [comment]
@everettraven I also tried to make the code compatible with the changes made in the versions of `kubectl` and `cli-runtime`. But I was not able to find a proper approach. I tried to make some changes regarding the changes in the func `Validator`. Could you please take a look https://github.com/NikhilSharmaWe/kubebuilder-declarative-pattern/commit/b2da05f0abcd5d05d24726e39228f4b5d540fc6e.

**Comment by everettraven** [comment]
@NikhilSharmaWe I made a draft PR so that I could have the tests run on any changes that I made. I was able to get the tests passing. You should be able to see the changes in the draft PR here:
https://github.com/kubernetes-sigs/kubebuilder-declarative-pattern/pull/222

**Comment by NikhilSharmaWe** [comment]
@everettraven Thanks for showing the correct approach for resolving the breaking changes.

- For func `NewFactory`: After seeing your approach I think my approach was almost correct but I needed to `resource.QueryParamFieldValidation` instead of `resource.QueryParamDryRun` in func  `NewQueryParamVerifier`.

- For func `NewApplyOptions`: In the future are we thinking to improve this approach by using the new functions added in the package `k8s.io/kubectl/pkg/cmd/apply` like https://pkg.go.dev/k8s.io/kubectl@v0.24.0/pkg/cmd/apply#ApplyFlags.ToOptions to produce the desired `*apply.ApplyOptions` instead of copying the function `NewApplyOptions` which was present in the pkg before. 
Or is this unnecessary?

File: pkg/patterns/declarative/pkg/applier/direct.go
Comment (everettraven) [inline pkg/patterns/declarative/pkg/applier/direct.go:152]: I would remove this `NewApplyOptions` function since we have a better implementation that goes along with the new changes in the k8s 1.24 api changes

**Review by everettraven — CHANGES_REQUESTED** [review]

File: pkg/patterns/declarative/pkg/applier/direct.go
Comment (everettraven) [inline pkg/patterns/declarative/pkg/applier/direct.go:48]: Nit: I would probably just remove this line completely rather than commenting it out.

**Review by everettraven — CHANGES_REQUESTED** [review]

**Review by everettraven — APPROVED** [review]
Looks good to me! Great job @NikhilSharmaWe ! Not sure I have permissions for it, but:

/lgtm
/approve

**Comment by NikhilSharmaWe** [comment]
@everettraven Thank you for considering the suggestion. Just wanted to ask how did found out that the **base name** should be `declarative-direct`.

**Comment by everettraven** [comment]
> @everettraven Thank you for considering the suggestion. Just wanted to ask how did found out that the **base name** should be `declarative-direct`.

It could be any string, but when looking into the details of the `apply.NewCmdApply` and `Apply.ToOptions` the *baseName* is used when printing messages or warnings (can be seen here: https://github.com/kubernetes/kubectl/blob/9a9633e4515045b45600d2e9cc5f0498c5a4d0d1/pkg/cmd/apply/apply.go#L301-L303).

I felt like `declarative-direct` was descriptive enough to point it to this repo and being in the `direct.go` file that the setup is being done in.

Now that I think about it, `kubebuilder-declarative-applier-direct` would be much more descriptive so feel free to change it to that if you would like, but I think the former is still rather descriptive.

**Review by camilamacedo86 — APPROVED** [review]
Great 👍 
Nice contribution @NikhilSharmaWe and @everettraven 🥇 

We need to ping the Owners of this repo so that they can help us move forward. 

@justinsb

**Comment by atoato88** [comment]
/lgtm

@justinsb 
Could you check this?

**Comment by everettraven** [comment]
@justinsb could you take a look at this?

We have a couple projects relying on this dependency to be upgraded to support k8s 1.24 before they can be upgraded. 

Thanks!

**Comment by everettraven** [comment]
@atoato88 or @justinsb is there anything else that needs to be done to get this merged?

**Comment by atoato88** [comment]
/approve
Sorry for my delay.
Thank you to create and update PR.

It already looks good for me, and I thinked it is also good to get review from other approver if possible.
But I approve this because this PR is getting time.

**Comment by jcanseco** [comment]
FYI we started to encounter an issue that seems to have been introduced in this PR: https://github.com/kubernetes-sigs/kubebuilder-declarative-pattern/issues/224

Wondering if anyone can take a look.

**Pull request merged**
